fdb-extensions/src/test/java/com/apple/foundationdb/async/RangeSetTest.java

```diff
@@ -233,66 +240,182 @@ public void insert() {
+                    .asList().join().size());
+            assertEquals(1, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x01, 0x01}, new byte[]{0x01, 0x02}))
+                    .asList().join().size());
+            assertEquals(1, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x01, 0x01}, new byte[]{0x02}))
+                    .asList().join().size());
+            assertEquals(1, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x04}, new byte[]{0x05}))
+                    .asList().join().size());
+            assertEquals(2, rs.missingRanges(multi.get(currTr++).snapshot()) // Should succeed if read at snapshot
+                    .asList().join().size());
+            assertFalse(rs.isEmpty(multi.get(currTr++).snapshot()).join());
+
+            // Should fail
+            final int failAfter = currTr;
+            assertFalse(rs.contains(multi.get(currTr++), new byte[]{0x03}).join());
+            assertFalse(rs.contains(multi.get(currTr++), new byte[]{0x03, 0x01}).join());
+            assertTrue(rs.contains(multi.get(currTr++), new byte[]{0x02}).join()); // fails because the key 0x02 itself is updated
+            assertEquals(2, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x01}, new byte[]{0x05}))
+                    .asList().join().size());
+            assertEquals(1, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x01}, new byte[]{0x02, 0x01}))
+                    .asList().join().size());
+            assertEquals(0, rs.missingRanges(multi.get(currTr++), r1)
+                    .asList().join().size());
+            assertEquals(1, rs.missingRanges(multi.get(currTr++), new Range(new byte[]{0x03, 0x01}, new byte[]{0x04, 0x01}))
+                    .asList().join().size());
+            assertFalse(rs.isEmpty(multi.get(currTr++)).join());
+
+            assertEquals(multi.size(), currTr);
+            multi.commit(IntStream.range(0, multi.size()).filter(i -> i >= failAfter).boxed().collect(Collectors.toSet()));
+        }
+    }
+
+    @Test
+    void failIfKeyInRangeIsInsertedConcurrently() {
+        try (MultipleTransactions multi = MultipleTransactions.create(db, 2)) {
+            Range r = new Range(new byte[]{0x01}, new byte[]{0x02});
+            Range overlappingRange = new Range(new byte[]{0x00, 0x01}, new byte[]{0x01, 0x01});
+
+            // Insert a range into the set
+            Transaction tr1 = multi.get(0);
+            assertTrue(rs.insertRange(tr1, overlappingRange, true).join());
+
+            // Read from a range that overlaps this range
+            Transaction tr2 = multi.get(1);
+            assertEquals(List.of(r), rs.missingRanges(tr2, r).asList().join());
+
+            // The second transaction should fail to commit
+            multi.commit(Set.of(1));
+        }
+    }
+
+    @Test
+    void insertNonOverlappingRanges() {
```

Comment:
It seems like it would be good to also have a test that does `[i, i+1)`, which would also be non-overlapping, but abutting. Or would that result in them merging, and thus conflicting?

Reply:
Added a test with `[i, i+1)`. This does succeed, because each one inserts a different key representing a different range. This means that it isn't coalesced, but given the goal here, it's probably better to have the lack of coalescence rather than to have transactions that fail.

In this test (and the others), I've added some additional tests to make sure that the set makes sense, i.e., there are gaps in the expected places